shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -1049,6 +1099,51 @@ - (FlutterPlatformViewController*)platformViewController {
 
 #pragma mark - Private methods
 
+- (void)addViewToEmbedderEngine:(FlutterViewController*)controller {
+  FlutterWindowMetricsEvent windowMetricsEvent{};
+  _populateMetricsEvent(controller, &windowMetricsEvent);
+  FlutterAddViewInfo info{
```

Comment:
As part of this change, could we add the blocking? I think you should be able to just copy-paste the Windows code:

https://github.com/flutter/engine/blob/8b77056c74d6349a641b07a767e9f7eabb66414d/shell/platform/windows/flutter_windows_engine.cc#L516-L550

Reply:
Blocking is much, much harder on macOS, because the macOS runner API creates an FVC and then link the FVC to the engine, instead of using the engine to create the FVC. This means that we can't simply delay returning the FVC until the addition is done, but have to record the state, block operations manually until the addition is done. I've tried it for like a week and it's definitely far from trivial. Due to the additional reason that the non-blocking way works fine and we don't plan to change the state in the forseeable future, it's not really worth the effort at least for this patch.

Reply:
If we don't block, the embedder and engine will have state that isn't synchronized for a short time. That's not necessarily bad, but we need to be careful around this.

On Windows, the biggest concern is if the platform thread removes a view while the raster thread is presenting to that view. Without blocking, this can crash on Windows. Is that a concern here?

Reply:
(Sorry for delaying) I don't think so, because presenting on macOS needs to happen on the platform thread anyway. I haven't observed such crash either. How often did you observe that crash on Windows without the blocking? (Also why doesn't locking and bailing resolve the crash?)